feat: configuration to disable sharing actions

kyrre opened the pull request:
## 📝 Summary

This adds an option to disable (hide) all the sharing functionality, e.g., "Share > Publish HTML to web". The issue is tracked in #5323 

![image](https://github.com/user-attachments/assets/b93fda55-ace7-43f0-b5a2-6e335f01eed9)
![image](https://github.com/user-attachments/assets/8714a450-9788-408b-9dc1-bddba3b84710)

## 🔍 Description of Changes

Introduces a new configuration setting "sharing", which is enabled by default. When it is disabled all the sharing related actions are hidden in the UI.

## 📋 Checklist

- [x] I have read the [contributor guidelines](https://github.com/marimo-team/marimo/blob/main/CONTRIBUTING.md).
- [x] For large changes, or changes that affect the public API: this change was discussed or approved through an issue, on [Discord](https://marimo.io/discord?ref=pr), or the community [discussions](https://github.com/marimo-team/marimo/discussions) (Please provide a link if applicable).
- [ ] I have added tests for the changes made.
- [x] I have run the code and verified that it works as expected.

## 📜 Reviewers

@mscolnick

kyrre commented:
Am I supposed to commit the OpenAPI code that is generated? 🤔

kyrre commented:
recheck

akshayka commented on marimo/_config/config.py:
```diff
@@ -381,6 +381,19 @@ class DatasourcesConfig(TypedDict):
+    - `enabled`: if `False`, sharing options will be hidden from the UI
+    """
+
+    enabled: bool
+
```
I think this could be more granular:

* html
* wasm

both booleans. wasm doesn't leak (we don't save code).

akshayka commented on frontend/src/components/app-config/user-config-form.tsx:
```diff
@@ -554,6 +554,41 @@ export const UserConfigForm: React.FC = () => {
                 )}
               />
             </SettingGroup>
+            <SettingGroup title="Sharing">
+              <FormField
+                control={form.control}
+                name="sharing.enabled"
+                render={({ field }) => (
+                  <div className="flex flex-col space-y-1">
+                    <FormItem className={formItemClasses}>
+                      <FormLabel className="font-normal">
+                        Enable sharing options
+                      </FormLabel>
+                      <FormControl>
+                        <Checkbox
+                          data-testid="sharing-enabled-checkbox"
+                          checked={field.value}
+                          disabled={field.disabled}
+                          onCheckedChange={(checked) => {
+                            field.onChange(Boolean(checked));
+                          }}
+                        />
+                      </FormControl>
+                      <FormMessage />
+                      <IsOverridden
+                        userConfig={config}
+                        name="sharing.enabled"
+                      />
+                    </FormItem>
+                    <FormDescription>
+                      When unchecked, sharing options (publish HTML, create
+                      WebAssembly link) will be hidden from the notebook actions
+                      menu.
+                    </FormDescription>
+                  </div>
+                )}
+              />
+            </SettingGroup>
```
I might remove this. If an org is disabling these feature for their team (via that `pyproject.toml`), likely you don't want to make it easy for developers to turn it back on in the UI.

kyrre replied:
> I might remove this. If an org is disabling these feature for their team (via that `pyproject.toml`), likely you don't want to make it easy for developers to turn it back on in the UI.

That's a good point, but would it even be possible for the user to do so? I was just reading the `pyproject.toml` setting in the [Configuration] documentation (https://docs.marimo.io/guides/configuration/#overriding-settings-with-pyprojecttoml) and it states that:

> Settings overridden in pyproject.toml or script metadata cannot be changed through the marimo editor's settings menu. Any changes made to overridden settings in the editor will not take effect.

But, anyway, it's not a very useful UI feature (how often would the user actually want to change it?) so I'll just remove it from the PR.

akshayka reviewed (changes requested):
Thanks for the PR! A couple of comments.

kyrre commented:
recheck

kyrre commented:
I have read the CLA Document and I hereby sign the CLA

kyrre commented:
> Premature approval, noticed a test is failing.
> 
> Also, if the sharing config is not required, then it shouldn't have a default value `config.py` / `config-schema.ts` I believe.

Sorry about that, I'm still getting familiar with the marimo codebase 😊

I've removed the default values from the config files. Since that means sharing.* can now be undefined, I added fallback logic in the UI: 

```js
// Fallback: if sharing is undefined, both are enabled by default
const sharingHtmlEnabled = resolvedConfig.sharing?.html ?? true;
const sharingWasmEnabled = resolvedConfig.sharing?.wasm ?? true;
```

Does this work for you, or did you have something else in mind? 🤔
